resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -309,6 +312,9 @@ private[spark] class ApplicationMaster(args: ApplicationMasterArguments) extends
         finish(FinalApplicationStatus.FAILED,
           ApplicationMaster.EXIT_UNCAUGHT_EXCEPTION,
           "Uncaught exception: " + StringUtils.stringifyException(e))
+    } finally {
+      metricsSystem.report()
```

Comment:
`metricsSystem` can be null at this point, can't it? in case of some issue during startup.

Reply:
Yes, better and more elegant to store metricSystem in an Option.